lfs/api.go

```diff
@@ -0,0 +1,93 @@
+package lfs
+
+import (
+	"io"
+	"sync"
+)
+
+// An abstract interface providing state & resource management for a specific Endpoint across
+// potentially multiple requests
+type ApiContext interface {
+	// Get the endpoint this context was constructed from
+	Endpoint() Endpoint
+	// Close the context & any resources it's using
+	Close() error
+
+	// Download a single object, return reader for data, size and any error
+	// Essentially the same as calling DownloadCheck() then DownloadObject()
+	Download(oid string) (io.ReadCloser, int64, *WrappedError)
+	// Check whether an object is available for download and return an object resource if so
+	DownloadCheck(oid string) (*ObjectResource, *WrappedError)
+	// Download a single object from an already identified resource from DownloadCheck(), return reader for data, size and any error
+	DownloadObject(obj *ObjectResource) (io.ReadCloser, int64, *WrappedError)
+	// Check whether an upload would be accepted for an object and return the resource to use if so
+	UploadCheck(oid string, sz int64) (*ObjectResource, *WrappedError)
+	// Perform the actual upload of an object having identified it will be accepted and the resource to use
+	UploadObject(o *ObjectResource, reader io.Reader) *WrappedError
+	// Perform a batch request for a number of objects to determine what can be uploaded/downloaded
+	Batch(objects []*ObjectResource) ([]*ObjectResource, *WrappedError)
```

Comment:
This interface isn't really ready to be extracted. We don't intend to maintain support for the current upload/download API once the batch API is finalized. I'm thinking an eventual interface would look something like this. 

```go
type ApiContext interface {
  Endpoint() Endpoint
  Close() error
  DownloadAll(objects []*ObjectResource)
  UploadAll(objects []*ObjectResource)
}
```

I think an `ApiContext` should be responsible for a single Git LFS operation. That includes a smudge filter that downloads a single object, a push command that's uploading 100 objects from the range of commits. I think that would remove the need to access cached `ApiContext` objects through `GetApiContext()`. I can't think of a reason you'd need to get an `ApiContext` for multiple endpoints, or multiple contexts for the same endpoint in a single Git LFS command.